### PR TITLE
introduce support for custom build outputs

### PR DESCRIPTION
**Pull request by ndeloof**
**What I did**
introduce `--output` flag on `build` command to export built artifact with alternate exporter.
AFAICT having this as part of the compose.yaml service definition doesn't make much sense as `compose up` scenario require image to be exported to container runtime local store. On the other hand, maybe there's room for new use-cases?

**Related issue**
fixes https://github.com/docker/compose/issues/9586

**(not mandatory) A picture of a cute animal, if possible in relation to what you did**

### CODE REVIEW

**Comment by Ryanf55** [comment]
What was the reason this was closed. I use build outputs to take unit tests results and report them back to CI. Without this support, looks like I cannot use compose in CI for build/test.

**Comment by israelglar** [comment]
> **What I did** introduce `--output` flag on `build` command to export built artifact with alternate exporter. AFAICT having this as part of the compose.yaml service definition doesn't make much sense as `compose up` scenario require image to be exported to container runtime local store. On the other hand, maybe there's room for new use-cases?
> 
> **Related issue** fixes #9586
> 
> **(not mandatory) A picture of a cute animal, if possible in relation to what you did**

I believe this should be reopened and merged. Maybe your reason for closing was that there was no "new use-cases", but as per @Ryanf55  and now also me, there are use cases where this options would be really helpful.

On my case I want our developer team and CI to use cache. For that I'm story my cache in an ECR using [cache_to:type=registry](https://docs.docker.com/build/cache/backends/registry/)  to be able to do that I need to use the [docker-container driver](https://docs.docker.com/build/builders/drivers/) .

This works for services that don't depend on anything else, but breaks when in use with [additional_contexts](https://docs.docker.com/reference/compose-file/build/#additional_contexts) referencing another service in compose. 
Because even tho the additional services are built and are locally present but not pushed, when the main service wants to load their metadata it looks for it in the registry but since it wasn't pushed yet (this is related to an open issue in [docker-container where it doesn't load local images](https://github.com/moby/buildkit/issues/2343) )

To summarise: 
 - I need to user `docker-container` to be able to export cache
 - `docker-container has a bug where it doesnt load local images
 
The really easy way to fix this is to push images as soon as they are built. In my testing if I use the buildx option `--output type=registry`  [docs here](https://docs.docker.com/reference/cli/docker/buildx/build/#registry)
 it fixes this problem because it pushes after building.
 
 That together with the open issue #9586 where many other people also want this feature in, I think we should reopen it and merge.

**Comment by israelglar** [comment]
Hi!
Thanks for answering so quickly, it always good to have project authors/contributors that are active and on top of stuff!

You're right. I was able to get this exact issue solved by using bake with the [x-bake key](https://docs.docker.com/build/bake/compose-file/#extension-field-with-x-bake)

In the future I can see how having this tight integration and delegating build to bake would be really good.

Thanks!
